pkg/controller/discovery/auth/sar.go

```diff
@@ -0,0 +1,331 @@
+package auth
+
+import (
+	"context"
+	"github.com/konveyor/mig-controller/pkg/controller/discovery/model"
+	"github.com/konveyor/mig-controller/pkg/logging"
+	project "github.com/openshift/api/project/v1"
+	auth "k8s.io/api/authentication/v1beta1"
+	"k8s.io/api/authorization/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/pointer"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+//
+// Shared logger.
+var Log *logging.Logger
+
+//
+// k8s Resources.
+const (
+	ALL       = "*"
+	Namespace = "namespaces"
+	PV        = "persistentvolumes"
+	PVC       = "persistentvolumeclaims"
+	Service   = "services"
+	Pod       = "pods"
+	PodLog    = "pods/log"
+)
+
+//
+// Verbs
+const (
+	ANY    = "*"
+	LIST   = "list"
+	GET    = "get"
+	CREATE = "create"
+	DELETE = "delete"
+	PATCH  = "patch"
+	UPDATE = "update"
+)
+
+//
+// SAR request.
+type Review = v1.ResourceAttributes
+
+//
+// Review Matrix.
+type Matrix struct {
+	// API group.
+	Group string
+	// Namespace
+	Namespace string
+	// Resource name.
+	Name string
+	// Resources.
+	Resources []string
+	// Verbs.
+	Verbs []string
+}
+
+//
+// Provides RBAC authorization.
+type RBAC struct {
+	client.Client
+	// Cluster
+	Cluster *model.Cluster
+	// The subject bearer token.
+	Token string
+	// Token review status.
+	tokenStatus auth.TokenReviewStatus
+	// Token authenticated.
+	authenticated *bool
+	// Allowed project list.
+	allowedProject map[string]bool
+}
+
+//
+// Allow/deny the subject review request.
+// Special (optimized) steps for namespaces:
```

Comment:
Summarizing the `Allow()` method in my own words for understanding. 

**For namespaces**:
 - Checks if token has `LIST` and `GET` permission on namespace using cached data
   - Grants access to project if token has both of these , returns true
 - If token doesn't have `LIST`, `GET` runs SAR check (due to `fallthrough`) which doesn't rely on cached data

**For any other kind of resource**
 - Immediately runs SAR check

---

I'm curious, are SAR checks inherently slow for namespaces moreso than other resources?